### ISSUE TITLE
Advanced KSampler Noise Injection Enhancement

### ISSUE THREAD

**Issue by brucew4yn3rp** [comfyanonymous/ComfyUI#6924]
### Feature Idea

When using the add_noise setting on a ksampler, the code uses the comfy.prepare_noise function.

The current prepare_noise function generates random noise for a latent image based on a given seed. While this provides basic noise generation functionality, it lacks flexibility in several key areas, such as controlling the intensity of the noise and applying more precise manipulation to the latent tensor. Specifically, the following limitations exist:

No Control Over Noise Intensity: The noise generated by the prepare_noise function is not scaled or adjusted, which could lead to unpredictable or excessive noise levels, especially if you need finer control over how much noise is injected into the latent representation.
> This especially causes issues with img2img when the add_noise is enabled and the start_at_step is greater than 0 (in an attempt to lower denoise)

Lack of Normalization: There is no option to normalize the noise to match the latent vector’s mean and standard deviation, which can cause the noise to feel disjointed or overly disruptive when added to the latent tensor.

**The Inject Latent Noise node allows for the introduction of latent noise in a more elegant manner. I propose modifying the prepare_noise function to incorporate the following improvements:**

**Noise Scaling:**
Introduce a noise_strength parameter to the advanced ksampler to scale the noise intensity, giving users more control over how much noise is added to the latent tensor.

**Noise Normalization:**
Implement an option to normalize the noise based on the latent tensor’s mean and standard deviation, ensuring that the noise blends seamlessly with the latent representation.

**Selective Noise Application:**
If a noise_mask (optional) is provided to the advanced ksampler, apply noise selectively to the masked areas. This allows for more targeted noise application without affecting the entire latent tensor.

### Existing Solutions
https://github.com/cubiq/ComfyUI_essentials/blob/main/sampling.py#L156
```
class InjectLatentNoise:
    @classmethod
    def INPUT_TYPES(s):
        return {"required": {
                    "latent": ("LATENT", ),
                    "noise_seed": ("INT", {"default": 0, "min": 0, "max": 0xffffffffffffffff}),
                    "noise_strength": ("FLOAT", {"default": 1.0, "min": -20.0, "max": 20.0, "step":0.01, "round": 0.01}),
                    "normalize": (["false", "true"], {"default": "false"}),
                },
                "optional": {
                    "mask": ("MASK", ),
                }}

    RETURN_TYPES = ("LATENT",)
    FUNCTION = "execute"
    CATEGORY = "essentials/sampling"

    def execute(self, latent, noise_seed, noise_strength, normalize="false", mask=None):
        torch.manual_seed(noise_seed)
        noise_latent = latent.copy()
        original_samples = noise_latent["samples"].clone()
        random_noise = torch.randn_like(original_samples)

        if normalize == "true":
            mean = original_samples.mean()
            std = original_samples.std()
            random_noise = random_noise * std + mean

        random_noise = original_samples + random_noise * noise_strength

        if mask is not None:
            mask = F.interpolate(mask.reshape((-1, 1, mask.shape[-2], mask.shape[-1])), size=(random_noise.shape[2], random_noise.shape[3]), mode="bilinear")
            mask = mask.expand((-1,random_noise.shape[1],-1,-1)).clamp(0.0, 1.0)
            if mask.shape[0] < random_noise.shape[0]:
                mask = mask.repeat((random_noise.shape[0] -1) // mask.shape[0] + 1, 1, 1, 1)[:random_noise.shape[0]]
            elif mask.shape[0] > random_noise.shape[0]:
                mask = mask[:random_noise.shape[0]]
            random_noise = mask * random_noise + (1-mask) * original_samples

        noise_latent["samples"] = random_noise

        return (noise_latent, )
```

### Other

Sample Integration:

```
def prepare_noise(latent_image, seed, noise_strength=1.0, normalize=False, mask=None, noise_inds=None):
    """
    Creates random noise based on a latent image, with options for scaling, normalization, and selective noise application.
    
    Parameters:
    - latent_image (Tensor): The latent image to which noise will be added.
    - seed (int): The random seed for noise generation.
    - noise_strength (float): Controls the intensity of the noise to be added. Default is 1.0.
    - normalize (bool): If True, normalizes the noise to match the latent image's statistics. Default is False.
    - mask (Tensor, optional): A binary mask to apply noise selectively. Default is None.
    - noise_inds (list, optional): Specific indices where noise should be applied. Default is None.
    
    Returns:
    - Tensor: The latent image with noise added.
    """
    
    generator = torch.manual_seed(seed)
    
    # Generate the initial random noise
    noise = torch.randn(latent_image.size(), dtype=latent_image.dtype, device=latent_image.device, generator=generator)
    
    # Normalize the noise to match the latent image's mean and std if required
    if normalize:
        mean = latent_image.mean()
        std = latent_image.std()
        noise = noise * std + mean
    
    # Scale the noise by noise_strength
    noise = noise * noise_strength
    
    # Apply the noise selectively if a mask is provided
    if mask is not None:
        mask = F.interpolate(mask.unsqueeze(1), size=(noise.shape[2], noise.shape[3]), mode="bilinear").squeeze(1)
        noise = noise * mask.unsqueeze(1) + latent_image * (1 - mask.unsqueeze(1))
    
    # If noise_inds are specified, apply different noise for each index
    if noise_inds is not None:
        unique_inds, inverse = np.unique(noise_inds, return_inverse=True)
        noises = []
        for i in range(unique_inds[-1] + 1):
            noise_part = torch.randn([1] + list(latent_image.size())[1:], dtype=latent_image.dtype, device=latent_image.device, generator=generator)
            if i in unique_inds:
                noises.append(noise_part)
        noises = [noises[i] for i in inverse]
        noise = torch.cat(noises, axis=0)

    # Add the generated noise to the latent image
    noisy_latent_image = latent_image + noise
    
    return noisy_latent_image
```

**Comment by ltdrdata** [comment]
I created this to allow noise to be input in a latent form.  
https://raw.githubusercontent.com/ltdrdata/ComfyUI-extension-tutorials/refs/heads/Main/ComfyUI-Inspire-Pack/workflow/seed-explorer.png)](https://raw.githubusercontent.com/ltdrdata/ComfyUI-extension-tutorials/refs/heads/Main/ComfyUI-Inspire-Pack/workflow/seed-explorer.png

And now, since the built-in `SamplerCustomAdvanced` allows for fine-grained control, you can also create and use a custom noise generator as needed.
![Image](https://github.com/user-attachments/assets/95d4b212-8549-48e7-a400-e6139dc9e35d)

**Comment by brucew4yn3rp** [comment]
Appreciate the response. My original query was more about the Advanced KSampler and why the control of noise seems to be neglected there. Sure, there's a workaround to that as well - just inject noise / manipulate the latent before passing it to the KSampler but I'd imagine that the Advanced KSampler should have an *advanced* built-in functionality for noise as well.

**Comment by ltdrdata** [comment]
KSampler and KSamplerAdvanced were initially designed as nodes to ensure they weren't too complex for end-users. 
(This was a design decision made at the time, and it cannot be changed now.)

Since then, nodes like SamplerCustom/SamplerCustomAdvanced have been added to meet the needs of power users who want more low-level control.

It is indeed more primitive, and precisely because it is more primitive, it allows for a wider range of control.

ComfyUI core provides both high-level and low-level nodes, so I believe that the mid-level functionality should now be handled by custom nodes.

**Comment by brucew4yn3rp** [comment]
I see, I wasn’t aware that even the Advanced KSampler was made with the goal of keeping the complexity low. Thanks for the clarification - I’ll close the feature request now.
